File: library_core/src/main/java/cn/flyfun/ktx/gamesdk/core/ui/dialog/PrivacyDialog.java
package cn.flyfun.ktx.gamesdk.core.ui.dialog;

import android.annotation.SuppressLint;
import android.app.Dialog;
import android.content.Context;
import android.graphics.Color;
import android.graphics.drawable.ColorDrawable;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.Window;
import android.view.WindowManager;
import android.webkit.CookieSyncManager;
import android.webkit.WebSettings;
import android.webkit.WebView;
import android.widget.Button;
import android.widget.FrameLayout;
import android.widget.ImageView;

import cn.flyfun.support.ResUtils;
import cn.flyfun.support.device.DeviceInfoUtils;

public class PrivacyDialog extends Dialog implements View.OnClickListener {


    private static final int ACTION_RETURN = 1000;
    private static final int ACTION_CONFIRM = 1001;

    private String url;

    private ImageView ivReturn;
    private Button btnConfirm;

    public PrivacyDialog(Context context, String url) {
        super(context);
        this.url = url;
        initView(context);
    }

    private void initView(Context context) {
        setCanceledOnTouchOutside(false);
        getWindow().setBackgroundDrawable(new ColorDrawable(Color.TRANSPARENT));
        requestWindowFeature(Window.FEATURE_NO_TITLE);

        View view = LayoutInflater.from(context).inflate(ResUtils.getResId(context, "ffg_privacy_dialog", "layout"), null);
        setContentView(view);

        WindowManager.LayoutParams attr = getWindow().getAttributes();
        if (attr != null) {
            //设置dialog 在布局中的位置
            attr.height = ViewGroup.LayoutParams.WRAP_CONTENT;
            attr.width = ViewGroup.LayoutParams.WRAP_CONTENT;
            attr.gravity = Gravity.CENTER;
        }

        ivReturn = findViewById(ResUtils.getResId(context, "ffg_iv_return", "id"));
        btnConfirm = findViewById(ResUtils.getResId(context, "ffg_btn_confirm", "id"));

        initWebView(context);

        initViewAction();
        initViewListener();
    }

    private void initViewAction() {
        ivReturn.setTag(ACTION_RETURN);
        btnConfirm.setTag(ACTION_CONFIRM);

    }

    private void initViewListener() {
        ivReturn.setOnClickListener(this);
        btnConfirm.setOnClickListener(this);
    }

    @SuppressLint({"SetJavaScriptEnabled", "NewApi"})
    private void initWebView(Context context) {
        WebView webView = new WebView(context);
        webView.setVerticalScrollBarEnabled(true);
        webView.setScrollBarStyle(View.SCROLLBARS_INSIDE_OVERLAY);
        FrameLayout webViewContainer = findViewById(ResUtils.getResId(context, "ffg_webview_container", "id"));
        webViewContainer.addView(webView);

        WebSettings webSetting = webView.getSettings();
        webSetting.setAllowFileAccess(true);
        webSetting.setLayoutAlgorithm(WebSettings.LayoutAlgorithm.NARROW_COLUMNS);
        webSetting.setSupportZoom(true);
        webSetting.setSaveFormData(true);
        webSetting.setSavePassword(true);
        webSetting.setBuiltInZoomControls(false);
        webSetting.setUseWideViewPort(true);
        webSetting.setSupportMultipleWindows(false);
        webSetting.setAppCacheEnabled(true);
        webSetting.setDomStorageEnabled(true);
        webSetting.setJavaScriptEnabled(true);
        webSetting.setGeolocationEnabled(true);
        webSetting.setAppCacheMaxSize(Long.MAX_VALUE);
        webSetting.setAppCachePath(context.getDir("appcache", 0).getPath());
        webSetting.setDatabasePath(context.getDir("databases", 0).getPath());
        webSetting.setGeolocationDatabasePath(context.getDir("geolocation", 0).getPath());
        webSetting.setPluginState(WebSettings.PluginState.ON_DEMAND);
        CookieSyncManager.createInstance(context);
        CookieSyncManager.getInstance().sync();

        if (DeviceInfoUtils.isNetworkConnected(context)) {
            webView.loadUrl(url);
        } else {
            //这种写法可以正确解码
            webView.loadData("网络异常,请检查重试", "text/html; charset=UTF-8", null);
        }
    }

    @Override
    public void onClick(View v) {
        if (v.getTag() != null) {
            switch ((Integer) v.getTag()) {
                case ACTION_RETURN:
                case ACTION_CONFIRM:
                    dismiss();
                    break;
            }
        }
    }
}

File: library_core/src/main/java/cn/flyfun/ktx/gamesdk/core/ui/dialog/PrivacyDialog.kt
package cn.flyfun.ktx.gamesdk.core.ui.dialog

import android.annotation.SuppressLint
import android.app.Dialog
import android.content.Context
import android.graphics.Color
import android.graphics.drawable.ColorDrawable
import android.view.*
import android.webkit.CookieSyncManager
import android.webkit.WebSettings
import android.webkit.WebView
import android.widget.Button
import android.widget.FrameLayout
import android.widget.ImageView
import cn.flyfun.ktx.gamesdk.core.entity.ClickType
import cn.flyfun.support.ResUtils
import cn.flyfun.support.device.DeviceInfoUtils


/**
 * @author #Suyghur.
 * Created on 2020/12/10
 */
class PrivacyDialog constructor(context: Context, val url: String) : Dialog(context), View.OnClickListener {

    private lateinit var ivReturn: ImageView
    private lateinit var btnConfirm: Button

    init {
        initView()
    }

    private fun initView() {
        setCanceledOnTouchOutside(false)
        window?.setBackgroundDrawable(ColorDrawable(Color.TRANSPARENT))
        requestWindowFeature(Window.FEATURE_NO_TITLE)

        val view = LayoutInflater.from(context).inflate(ResUtils.getResId(context, "ffg_privacy_dialog", "layout"), null)
        setContentView(view)

        val attr = window?.attributes as WindowManager.LayoutParams
        attr.height = ViewGroup.LayoutParams.WRAP_CONTENT
        attr.width = ViewGroup.LayoutParams.WRAP_CONTENT
        attr.gravity = Gravity.CENTER

        ivReturn = findViewById(ResUtils.getResId(context, "ffg_iv_return", "id"))
        btnConfirm = findViewById(ResUtils.getResId(context, "ffg_btn_confirm", "id"))

        initWebView()
        initViewAction()
        initViewListener()
    }

    @SuppressLint("SetJavaScriptEnabled", "NewApi")
    private fun initWebView() {
        val webView = WebView(context)
        webView.isVerticalScrollBarEnabled = true
        webView.scrollBarStyle = View.SCROLLBARS_INSIDE_OVERLAY
        val webViewContainer = findViewById<FrameLayout>(ResUtils.getResId(context, "ffg_webview_container", "id"))
        webViewContainer.addView(webView)

        val webSetting = webView.settings
        webSetting.allowFileAccess = true
        webSetting.layoutAlgorithm = WebSettings.LayoutAlgorithm.NARROW_COLUMNS
        webSetting.setSupportZoom(true)
        webSetting.saveFormData = true
        webSetting.savePassword = true
        webSetting.builtInZoomControls = false
        webSetting.useWideViewPort = true
        webSetting.setSupportMultipleWindows(false)
        webSetting.setAppCacheEnabled(true)
        webSetting.domStorageEnabled = true
        webSetting.javaScriptEnabled = true
        webSetting.setGeolocationEnabled(true)
        webSetting.setAppCacheMaxSize(Long.MAX_VALUE)
        webSetting.setAppCachePath(context.getDir("appcache", 0).path)
        webSetting.databasePath = context.getDir("databases", 0).path
        webSetting.setGeolocationDatabasePath(context.getDir("geolocation", 0).path)
        webSetting.pluginState = WebSettings.PluginState.ON_DEMAND
        CookieSyncManager.createInstance(context)
        CookieSyncManager.getInstance().sync()

        if (DeviceInfoUtils.isNetworkConnected(context)) {
            webView.loadUrl(url)
        } else {
            //这种写法可以正确解码
            webView.loadData("网络异常,请检查重试", "text/html; charset=UTF-8", null)
        }
    }

    private fun initViewAction() {
        ivReturn.tag = ClickType.ACTION_RETURN
        btnConfirm.tag = ClickType.ACTION_CONFIRM
    }

    private fun initViewListener() {
        ivReturn.setOnClickListener(this)
        btnConfirm.setOnClickListener(this)
    }

    override fun onClick(v: View?) {
        v?.apply {
            when (tag as Int) {
                ClickType.ACTION_RETURN, ClickType.ACTION_CONFIRM -> dismiss()
            }
        }
    }
}